esp/platform/espcontext.cpp

```diff
@@ -1018,3 +1019,31 @@ IEspServer* queryEspServer()
 {
     return dynamic_cast<IEspServer*>(getESPContainer());
 }
+
+IRemoteConnection* getSDSConnectionWithRetry(const char* xpath, unsigned mode, unsigned connectionTimeoutMS, int maxRetries, unsigned retrySeconds)
+{
+    int retries = 0;
+    while (true)
+    {
+        try
+        {
+            Owned<IRemoteConnection> conn = querySDS().connect(xpath, myProcessSession(), mode, connectionTimeoutMS);
+            if (!conn)
+                throw MakeStringException(-1, "Failed in SDS connect.");
+            return conn.getClear();
+        }
+        catch(IException* e)
+        {
+            if (retries == maxRetries)
+                return nullptr;
+
+            StringBuffer msg;
+            IERRLOG("getSDSConnectionWithRetry() Exception %d:%s. Will retry after %d seconds.",
+                e->errorCode(), e->errorMessage(msg).str(), retrySeconds);
+            e->Release();
+            sleep(retrySeconds);
```

Comment:
Normal pattern is to catch the timeout exception, and retry immediately since it has sat blocked/idle on connect up to the connect timeout already, there's no point a separate idle/sleep time.

And clearer with just an overall timelimit passed in rathar than retries/sleeps/per-connect-timeout..,
the per connect timeout could be passed in but is internal and less significant and unlikely to want to be customized.

I think method should look like this:
```
IRemoteConnection* getSDSConnectionWithRetry(const char* xpath, unsigned mode, unsigned timeoutMs)
{
    CTimeMon timer(timeoutMs);
    unsigned remaining;
    while (!timer.timedout(&remaining))
    {
        try
        {
            unsigned connTimeoutMs = remaining > SESSION_SDS_LOCK_TIMEOUT ? SESSION_SDS_LOCK_TIMEOUT : remaining;
            Owned<IRemoteConnection> conn = querySDS().connect(xpath, myProcessSession(), mode, connTimeoutMs);
            if (!conn)
                throw MakeStringException(-1, "getSDSConnectionWithRetry() : unabled to establish connection to : %s", xpath);
            return conn.getClear();
        }
        catch (ISDSException *e)
        {
            if (SDSExcpt_LockTimeout != e->errorCode())
                throw;
            IERRLOG(e, "getSDSConnectionWithRetry()");
            e->Release();
        }
    }
    return nullptr;
}
```